File: 2021/java/45.jump-game-ii.java
class Solution {
    public int jump(int[] nums) {
        if (nums == null || nums.length == 0) return 0;
        int len = nums.length;
        int[] dp = new int[len];
        Arrays.fill(dp, Integer.MAX_VALUE);
        dp[0] = 0;
        for (int i = 0; i < len; i++) {
            for (int j = i + 1; j <= i + nums[i] && j < len; j++) {
                dp[j] = Math.min(dp[j], dp[i] + 1);
            }
        }
        return dp[len - 1];
        
    }
}

File: 2021/python/45.jump-game-ii.py
class Solution:
    def jump(self, nums: List[int]) -> int:
        if not nums or len(nums) == 0:
            return 0
        dp = [float("inf") for _ in nums]
        dp[0] = 0
        for i in range(1, len(nums)):
            for j in range(i):
                if nums[j] >= i - j:
                    dp[i] = min(dp[i], dp[j] + 1)
        return dp[-1]
     
    def jump(self, nums: List[int]) -> int:
        if not nums or len(nums) == 0:
            return 0
        dp = [float("inf") for _ in nums]
        dp[0] = 0
        for i in range(1, len(nums)):
            dp[i] = min([dp[j] + 1 for j in range(i) if nums[j] >= i - j])
            
        return dp[-1]
    
